test/spmd/test_xla_distributed_checkpoint.py

```diff
@@ -0,0 +1,165 @@
+import os
+import tempfile
+import unittest
+import test_xla_sharding_base
+
+import torch
+import torch.distributed.checkpoint as dist_cp
+import torch_xla.core.xla_model as xm
+import torch_xla.runtime as xr
+import torch_xla.experimental.xla_sharding as xs
+
+from torch.distributed.checkpoint.default_planner import (
+    create_default_local_save_plan,
+    create_default_global_save_plan,
+)
+from torch_xla.experimental.distributed_checkpoint import SPMDLoadPlanner
+
+
+class DistributedCheckpointTestBase(test_xla_sharding_base.XlaShardingTest):
+
+  @classmethod
+  def setUpClass(cls):
+    os.environ['XLA_USE_SPMD'] = '1'
+    super().setUpClass()
+
+  def _get_sharded_model(self, mesh_shape=None):
+    # Return a sharded SimpleLinear model with fc1.weight sharded and
+    # fc2.weight explicitly replicated
+    mesh_shape = mesh_shape or (1, self.n_devices)
+    model = self.SimpleLinear().to(xm.xla_device())
+    mesh = self._get_mesh(mesh_shape)
+    xs.mark_sharding(model.fc1.weight, mesh, (0, 1))
+    xs.mark_sharding(model.fc2.weight, mesh, (None, None))
+    return model
+
+  def _get_default_local_metadata(self):
+    # Create a default Metadata instance for a SimpleLinear model by using
+    # the default local save plan on the model's CPU state_dict
+    cpu_state_dict = self.SimpleLinear().state_dict()
+    local_plan = create_default_local_save_plan(cpu_state_dict, True)
+    _, md = create_default_global_save_plan([local_plan])
+    return md
+
+  def _same_shard_data(self, shards, others) -> bool:
+    for a, b in zip(shards, others):
+      if not torch.allclose(a.data, b.data):
+        return False
+    return True
+
+
+class ReshardingTest(DistributedCheckpointTestBase):
+
+  def _save_and_restore(self,
+                        model_in,
+                        model_out,
+                        save_planner=None,
+                        load_planner=None):
+    """
+    Checkpoint model_in using the provided save_planner and load into model_out
+    using the provided load_planner, and assert model_out equals model_in after
+    the load. If either planner is not specified, the DefaultPlanner is used.
+    """
+    tmpdir = tempfile.mkdtemp()
+
+    # Save an unsharded model using the provided save planner
+    dist_cp.save_state_dict(
+        state_dict=model_in.state_dict(),
+        storage_writer=dist_cp.FileSystemWriter(tmpdir),
+        planner=save_planner,
+        no_dist=True,  # Single-host checkpoint doesn't require a process group
+    )
+
+    # Load the checkpoint using the provided load planner
+    for p1, p2 in zip(model_in.parameters(), model_out.parameters()):
+      self.assertFalse(torch.allclose(p1, p2))
+    dist_cp.load_state_dict(
```

Comment:
It looks like the LoadPlanner should be able to handle the resharding of the consolidated checkpoints. I wonder how does it handle the case where the save checkpoint has a different sharding spec than the loading model?

Reply:
Seems like it's handled by create_read_items_for_chunk_list?

Reply:
Then is there a simple formula to calculate the number of ReadItems? Or I have to refer to the create_read_items_for_chunk_list algorithm?